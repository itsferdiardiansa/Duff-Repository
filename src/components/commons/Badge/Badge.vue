<template>
  <div v-bind="$attrs" :class="customClass">
    <div :class="`${prefixClass}-badge--wrapper`" v-if="!dot">
      <template v-if="icon.length">
        <div :class="`${prefixClass}-badge--icon`">
          <font-awesome-icon :icon="icon" ref="iconEl" />
        </div>
      </template>

      <template v-if="hasSlot || label.length">
        <div
          :class="[`${prefixClass}-badge--label`, { 'has-icon': icon.length }]"
        >
          <slot>
            {{ label }}
          </slot>
        </div>
      </template>
    </div>
  </div>
</template>
<script>
import { computed, getCurrentInstance, ref } from 'vue';
import defaultProps from './props';

export default {
  name: 'CMBadge',
  props: defaultProps,
  setup(props, { slots }) {
    let root = getCurrentInstance();
    let iconEl = ref();

    const getVariantClass = prefixBtnClass => {
      const { inverse, variant } = props;
      let btnClass = `${prefixBtnClass}--${variant}`;

      if (inverse) btnClass += '-inverse';

      return btnClass;
    };

    const customClass = computed(() => {
      const {
        data: { prefixClass },
      } = root;
      const { size, rounded, bold, pill, dot } = props;
      const btnClass = prefixClass.concat('-badge');
      let customClass = [btnClass];

      customClass.push(getVariantClass(btnClass));

      if (size) customClass.push(btnClass.concat('--' + size));

      if (bold) customClass.push('text-bold');

      if (pill) customClass.push('pill');
      else if (!rounded) customClass.push('no-rounded');

      if (dot) customClass.push('dot');

      return [customClass.join(' ')];
    });

    const hasSlot = computed(() => !!slots.default);

    return {
      customClass,
      hasSlot,
      iconEl,
    };
  },
};
</script>
<style lang="scss">
@import './variant.scss';
@import './size.scss';

.#{$prefixClass}-badge {
  @apply h-9 focus:outline-none rounded px-3 transition duration-300 relative inline-block;
  @apply text-gray-800 overflow-hidden whitespace-nowrap;

  @include variant;

  @include size;

  &.no-rounded {
    @apply rounded-none;
  }

  &.pill {
    @apply rounded-full;
  }

  &.text-bold {
    @apply font-bold;
  }

  &.disabled {
    @apply disabled:opacity-70 pointer-events-none;
  }

  &--wrapper {
    @apply h-full flex-grow-0 flex flex-wrap items-center justify-center;
  }

  &--label {
    @apply pt-0.5;

    &.has-icon {
      @apply ml-2;
    }
  }
}
</style>
