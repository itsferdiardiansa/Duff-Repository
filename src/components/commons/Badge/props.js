export default {
  label: {
    type: [String, Number],
    default: '',
  },
  inverse: {
    type: Boolean,
    default: false,
  },
  variant: {
    type: String,
    default: 'light',
    validator: function (variant) {
      return ~[
        'primary',
        'danger',
        'warning',
        'dark',
        'success',
        'light',
        'orange',
      ].indexOf(variant);
    },
  },
  size: {
    type: String,
    default: 'xs',
    validator: function (variant) {
      return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant);
    },
  },
  icon: {
    type: Array,
    default: () => [],
  },
  bold: {
    type: Boolean,
    default: true,
  },
  rounded: {
    type: Boolean,
    default: true,
  },
  pill: {
    type: Boolean,
    default: false,
  },
  dot: {
    type: Boolean,
    default: false,
  },
};
