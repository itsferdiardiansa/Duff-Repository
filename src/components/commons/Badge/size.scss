@mixin size {
  &--xs {
    @apply h-6 text-xs px-2;
    
    &.dot {
      @apply h-4;
    }

    svg {
      @apply h-2.5;
    }
  }

  &--sm {
    @apply h-7 text-sm px-2.5;

    &.dot {
      @apply h-5;
    }

    svg {
      @apply h-3;
    }
  }

  &--base {
    @apply h-9 px-3;

    &.dot {
      @apply h-6;
    }

    svg {
      @apply h-3.5;
    }
  }

  &--lg {
    @apply h-12 text-lg px-4;

    &.dot {
      @apply h-8;
    }

    svg {
      @apply h-4;
    }
  }

  &--xl {
    @apply h-14 text-xl px-5;

    &.dot {
      @apply h-10;
    }

    svg {
      @apply h-5 mr-2;
    }
  }
}