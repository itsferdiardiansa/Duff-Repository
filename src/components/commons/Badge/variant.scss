@mixin variant {
  &.primary {
    @apply bg-primary text-gray-200;
  }

  &.danger {
    @apply bg-danger text-gray-200;
  }

  &.warning {
    @apply bg-warning text-gray-800;
  }

  &.success {
    @apply bg-success text-gray-200;
  }

  &.dark {
    @apply bg-dark text-gray-200;
  }
}
