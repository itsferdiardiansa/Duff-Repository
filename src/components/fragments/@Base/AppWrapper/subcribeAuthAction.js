export default (type, router) => {
  // Router push
  if (type === 'user/setToken') router.push('/');
};
