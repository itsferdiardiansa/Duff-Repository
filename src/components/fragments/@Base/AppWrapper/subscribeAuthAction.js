const alert = (content, variant) => sAlert.show({ content, variant });

export default (type, router) => {
  switch (type) {
    case 'user/loginSuccess':
      alert('Welcome aboard', 'success');
      break;
    case 'user/loginFailed':
      alert('Failed to login', 'danger');
      break;
    case 'user/removeAuth':
      alert('Thank you!', 'success');

      router.push('/login');
      break;
    default:
      break;
  }
};
