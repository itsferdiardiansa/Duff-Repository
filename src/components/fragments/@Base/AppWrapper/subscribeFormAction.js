export default (type, payload, router) => {
  if (
    ['form.create', 'form.update', 'form.delete'].includes(
      payload?.requestData?.action
    )
  ) {
    const variant = type.match(/fetchFailed/g) ? 'danger' : 'success';
    const content = payload?.responseData?.message;

    sAlert.show({ variant, content });

    if (
      ['form.create', 'form.update'].includes(payload?.requestData?.action) &&
      variant !== 'danger' &&
      Reflect.has(payload?.requestData, 'redirectUrl')
    ) {
      router.push(payload?.requestData.redirectUrl);
    }
  }
};
