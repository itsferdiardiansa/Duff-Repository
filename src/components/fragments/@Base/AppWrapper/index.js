/* eslint-disable no-unused-vars */
import {
  defineComponent,
  h,
  ref,
  getCurrentInstance,
  unref,
  computed,
} from 'vue'
import { useStore } from 'vuex'
import { useRouter } from 'vue-router'
import globalSetting from '@plugin/globalSetting'
import ErrorComponent from '@base/Error'
import subscribeFormAction from './subscribeFormAction'
import asyncModuleLoad from './asyncModuleLoad'

const AppWrapper = defineComponent({
  beforeCreate() {
    const {
      appContext: { app },
    } = getCurrentInstance()
    const store = useStore()
    const router = useRouter()

    store.subscribe(({ payload }, state) => {
      subscribeFormAction(payload, router)
    })

    // Handle async module load
    asyncModuleLoad()

    // register global mixin
    app.mixin(globalSetting)
  },
  beforeRouteEnter() {
    console.log('before router enter wrapper')
  },
  errorCaptured(error, component) {},
  setup(props, { slots }) {
    const store = useStore()

    const errorChunkModule = computed(() => store.getters['app/getChunks'])

    return () => {
      return unref(errorChunkModule).onError
        ? h(ErrorComponent, {})
        : slots.default()
    }
  },
})

export default AppWrapper
