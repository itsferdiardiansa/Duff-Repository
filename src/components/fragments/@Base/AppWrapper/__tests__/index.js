import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import AppWrapper from '@base/AppWrapper';
import Badge from '@common/Badge';
import appModule from '@store/modules/app';

describe('fragment/AppWrapper', () => {
  let Component, store, alertElement, prefixClass;

  beforeAll(() => {
    store = createStore({
      modules: {
        app: appModule,
      },
    });

    Component = mount(
      <AppWrapper>
        <Badge label="its_ferdi" variant="success" />
      </AppWrapper>,
      {
        global: {
          plugins: [store],
        },
      }
    );

    prefixClass = Component.rootVM.prefixClass;
  });

  it('will render children and match snapshot', () => {
    expect(Component.html()).toMatchSnapshot();
  });

  it('will render loading when progress fetch chunks', async () => {
    await Component.vm.$store.dispatch('app/progressChunkLoad');

    expect(Component.find('h1').text()).toEqual('Please wait');
    expect(Component.html()).toMatchSnapshot();
  });

  it('will render error when progress fetch chunks fail', async () => {
    await Component.vm.$store.dispatch('app/progressChunkFailed');

    expect(Component.find('h1').text()).toEqual(
      'Sorry there is a problem with our system'
    );
    expect(Component.find('p').text()).toEqual(
      'Wait a minute, let me clean up. Please try again or come back later.'
    );
    expect(Component.html()).toMatchSnapshot();
  });
});
