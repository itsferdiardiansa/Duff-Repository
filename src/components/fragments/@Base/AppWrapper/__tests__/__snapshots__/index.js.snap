// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment/AppWrapper will render children and match snapshot 1`] = `
<div class="ss-badge ss-badge--success ss-badge--xs text-bold">
  <div class="ss-badge--wrapper">
    <!--v-if-->
    <div class="ss-badge--label">its_ferdi</div>
  </div>
</div>
`;

exports[`fragment/AppWrapper will render error when progress fetch chunks fail 1`] = `
<div class="ss-error fade-slide-enter-from fade-slide-enter-active">
  <div class="ss-error--img"><img svg-inline=""></div>
  <div class="ss-error--description">
    <div class="message">
      <h1>Sorry there is a problem with our system</h1>
      <p> Wait a minute, let me clean up. Please try again or come back later. </p>
    </div>
    <div class="action-button"><button class="ss-button ss-button--dark ss-button--xl text-bold">
        <div class="ss-button--wrapper">
          <!--v-if-->
          <div class="ss-button--label">Reload</div>
        </div>
      </button></div>
  </div>
</div>
`;

exports[`fragment/AppWrapper will render loading when progress fetch chunks 1`] = `
<div class="ss-loading fade-enter-from fade-enter-active">
  <div class="anim">
    <div></div>
    <div></div>
    <div></div>
  </div>
  <h1>Please wait</h1>
</div>
`;
